### DISCUSSION TITLE
Array.at 不能polyfill

### DISCUSSION THREAD

**Discussion by atguo** [umijs/umi#11972] [Q&A] [answered]
无论是在umi3还是umi4中，array.at都没有被添加polyfill
我配置了
`
{
targets: { chrome: 49, firefox: 64, safari: 10, edge: 13, ios: 10 }}
`

Array.at兼容性如下
![image](https://github.com/umijs/umi/assets/13193919/1ba3c76c-580a-449a-8f7f-c5c3785fe199)

测试方式
1. 通过yarn create umi创建项目
2. 在pages/index.tsx中添加代码： 
`console.log([1,2,3].at(-2));`
4. yarn build

在构建产物中
![image](https://github.com/umijs/umi/assets/13193919/e1bc1247-721b-4a2d-98b6-f60cefbb09d4)

我的理解是，配置了targets以后，构建产物是不是应该将代码中，不符合targets的语言特性自动添加polyfill，将其编译为低版本浏览器可以运行的代码，不知道是我的理解有问题，还是这是个bug

**Comment by MissNanLan** [comment]
我试了下 es 2021 可以转换，如optional chaining ?.  、??
但是es 2022 不可以， 如Array.at()、hasOwnProperty

<img width="1083" alt="image" src="https://github.com/umijs/umi/assets/22790863/40a7e461-846c-46bf-a700-78837383d3cd">

不会是babel-loader的缘故吧

**Comment by atguo** [comment]
看了下项目node_modules
umi3的项目中，core-js版本是3.6.5，4年前的版本，那时候Array.at应该还没诞生(或者是没到stage3), 所以这个版本的core-js没有办法polyfill
umi4里可能polyfill了，看了下core-js的实现，没有细看，大概感觉好像是改了原型链来实现的，所以build之后的产物中可以调用.at。 但是我找不到低版本浏览器验证

**Reply by atguo** [reply]
问题来了，我能在umi3中换core-js的版本吗？
是直接在package中装个最新的可以吗？

**Reply by MissNanLan** [reply]
试试呗， 安装最新的core-js
在entry  import

或者试试overriders

**Comment by xiaohuoni** [comment, marked as answer]
真实环境中不能跑？
`?. 、??` 这些是语法，会被转化。
`at` 是方法，通过补丁，产物中存在，是合理的吧？？

**Reply by atguo** [reply]
是的，现在搞懂了。
问题实际上是并没有Array.at的补丁(因为core-js版本较低)
导致构建后在某些低版本浏览器报错了

**Reply by atguo** [reply]
之前以为 .at之类的语言特性，通过编译之类的方法，吧代码转换为低版本可用的
主要还是不懂polyfill，babel之间的关系

**Comment by fz6m** [comment]
有没有某个特性的 polyfill 直接看 `src/.umi/core/polyfill.ts` 就行了：

```ts
import ".../node_modules/core-js/modules/es.array.at.js";
```

如果你发现有，就是有的，如果你发现没有，那就是没有这个 polyfill 。

默认情况下，包含了你设定的 targets 到最新 esnext 所有的 polyfill 。

至于你在产物里搜，他不一定是 polyfill 到局部还是原型链上了，你可能要搜 `array.prototype` 、`prototype.at` 之类的，不方便，直接看 `polyfill.ts` 吧。
